---
title: 'Working with genomes in Python: finding 3-mers'
author: Yuwan Guo
date: '2020-05-16'
slug: working-with-genomes-in-python-finding-3-mers
categories: []
tags: []
description: ''
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, fig.align = "center", warning = F, message = F,
tidy=TRUE, tidy.opts=list(width.cutoff=60), R.options=list(max.print=100))
```

In bioinformatics, k-mers refer to all the possible subsequences (of length k) from a read obtained through DNA sequencing. For example, if the DNA sequencing read is "ATCATCATG", then the 3-mers in that read include "ATC" (which occurs twice), "TCA" (which occurs twice), "CAT" (occurs twice), and "ATG" (occurs once).

We can write a function in Python that takes a string of nucleotides as input and returns a dictionary with all 3-mers present in that string, and the number of times that each 3-mer occurs. The function written could be validated by finding the 3-mers in the DNA sequence my_seq1 and my_seq2 defined below.

```{python}
my_seq1 = "ATCATCATG"
my_seq2 = "CAGCCCAATCAGGCTCTACTGCCACTAAACTTACGCAGGATATATTTACGCCGACGTACT"

def three_mer(sequence): #argument will be string object
    newlist = [] #new list
    for i in range(len(sequence)-2):
        newlist.append(sequence[i]+sequence[i+1]+sequence[i+2])
    counts = {} #new dict
    for c in newlist: #for each 3-mer in sequence
        if c in counts: #have we seen it before in this sequence?
            counts[c]+=1 #if yes, increase count by 1
        else:
            counts[c]=1 #otherwise, set count
    return counts
    
three_mer(my_seq1)
```

```{python}
three_mer(my_seq2)
```