---
title: 'Working with genomes in Python: finding 3-mers'
author: Yuwan Guo
date: '2020-05-14'
slug: working-with-genomes-in-python-finding-3-mers
categories: []
tags: []
description: ''
---



<p>In bioinformatics, k-mers refer to all the possible subsequences (of length k) from a read obtained through DNA sequencing. For example, if the DNA sequencing read is “ATCATCATG”, then the 3-mers in that read include “ATC” (which occurs twice), “TCA” (which occurs twice), “CAT” (occurs twice), and “ATG” (occurs once).</p>
<p>We can write a function in Python that takes a string of nucleotides as input and returns a dictionary with all 3-mers present in that string, and the number of times that each 3-mer occurs. The function written could be validated by finding the 3-mers in the DNA sequence my_seq1 and my_seq2 defined below.</p>
<pre class="python"><code>my_seq1 = &quot;ATCATCATG&quot;
my_seq2 = &quot;CAGCCCAATCAGGCTCTACTGCCACTAAACTTACGCAGGATATATTTACGCCGACGTACT&quot;

def three_mer(sequence): #argument will be string object
    newlist = [] #new list
    for i in range(len(sequence)-2):
        newlist.append(sequence[i]+sequence[i+1]+sequence[i+2])
    counts = {} #new dict
    for c in newlist: #for each 3-mer in sequence
        if c in counts: #have we seen it before in this sequence?
            counts[c]+=1 #if yes, increase count by 1
        else:
            counts[c]=1 #otherwise, set count
    return counts
    
three_mer(my_seq1)</code></pre>
<pre><code>## {&#39;ATC&#39;: 2, &#39;TCA&#39;: 2, &#39;CAT&#39;: 2, &#39;ATG&#39;: 1}</code></pre>
<pre class="python"><code>three_mer(my_seq2)</code></pre>
<pre><code>## {&#39;CAG&#39;: 3, &#39;AGC&#39;: 1, &#39;GCC&#39;: 3, &#39;CCC&#39;: 1, &#39;CCA&#39;: 2, &#39;CAA&#39;: 1, &#39;AAT&#39;: 1, &#39;ATC&#39;: 1, &#39;TCA&#39;: 1, &#39;AGG&#39;: 2, &#39;GGC&#39;: 1, &#39;GCT&#39;: 1, &#39;CTC&#39;: 1, &#39;TCT&#39;: 1, &#39;CTA&#39;: 2, &#39;TAC&#39;: 4, &#39;ACT&#39;: 4, &#39;CTG&#39;: 1, &#39;TGC&#39;: 1, &#39;CAC&#39;: 1, &#39;TAA&#39;: 1, &#39;AAA&#39;: 1, &#39;AAC&#39;: 1, &#39;CTT&#39;: 1, &#39;TTA&#39;: 2, &#39;ACG&#39;: 3, &#39;CGC&#39;: 2, &#39;GCA&#39;: 1, &#39;GGA&#39;: 1, &#39;GAT&#39;: 1, &#39;ATA&#39;: 2, &#39;TAT&#39;: 2, &#39;ATT&#39;: 1, &#39;TTT&#39;: 1, &#39;CCG&#39;: 1, &#39;CGA&#39;: 1, &#39;GAC&#39;: 1, &#39;CGT&#39;: 1, &#39;GTA&#39;: 1}</code></pre>
