---
title: 'Working with genomes in Python: finding 3 mers'
author: Yuwan Guo
date: '2020-05-16'
slug: working-with-genomes-in-python-finding-3-mers
categories: []
tags: []
description: ''
---



<p>In bioinformatics, k-mers refer to all the possible subsequences (of length k) from a read obtained through DNA sequencing. For example, if the DNA sequencing read is “ATCATCATG”, then the 3-mers in that read include “ATC” (which occurs twice), “TCA” (which occurs twice), “CAT” (occurs twice), and “ATG” (occurs once).</p>
<p>We can write a function in Python that takes a string of nucleotides as input and returns a dictionary with all 3-mers present in that string, and the number of times that each 3-mer occurs. The function written could be validated by finding the 3-mers in the DNA sequence my_seq1 and my_seq2 defined below.</p>
<pre class="python"><code>mouse_histone = &quot;MARTKQTARKSTGGKAPRKQLATKAARKSAPATGGVKKPHRYRPLTVALREIRRYQKSTELLIRKLPFQRLVREIAQDFKTDLRFQSSAVMALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA&quot;
human_histone = &quot;MARTKQTARKSTGGKAPRKQLATKAQRKSARATGGVKKPHRYRPGTVALREIRRYQKSTELLIRKLPFQRLVTEIAQDFKTDLRFQSSAVNALQEACEAYLVGLFEDTNLCAIHAKRVTIMPKDIQLARRIRGERA&quot;

def seq_identity (seq1, seq2):
    count = 0
    if len(seq1) == len(seq2):
        for i in range(len(seq1)):
            if seq1[i] == seq2[i]:
                count += 1
        percent = count/len(seq1)*100
        percentstring = str(round(percent, 2))
        print(&#39;Pairwise sequence identity is:&#39;, percentstring, &#39;%&#39;)
    else:
        print(&#39;ERROR: the two sequences are of different length!&#39;)
    
seq_identity(mouse_histone, human_histone)</code></pre>
<pre><code>## Pairwise sequence identity is: 96.32 %</code></pre>
